srcpkgs/dmd/files/musl.patch

```diff
@@ -0,0 +1,21 @@
+--- dmd/src/dmd/link.d
++++ dmd/src/dmd/link.d
+@@ -759,8 +759,7 @@
+             perror("unable to create pipe to linker");
+             return -1;
+         }
+-        // vfork instead of fork to avoid https://issues.dlang.org/show_bug.cgi?id=21089
+-        childpid = vfork();
++        childpid = fork();
+         if (childpid == 0)
+         {
+             // pipe linker stderr to fds[0]
+@@ -768,7 +767,7 @@
+             close(fds[0]);
+             execvp(argv[0], argv.tdata());
+             perror(argv[0]); // failed to execute
+-            _exit(-1);
++            return -1;
```

Comment:
Why is a return here? Using a `return` feels very wrong.... I don't think I would revert this line.

`vfork` is available on musl, they are probably failing to `#define _BSD_SOURCE` at the top of the file, or whatever the D equivalent is.

Reply:
D does not use C headers. I tried to track down the declaration of `vfork` in glibc druntime, which led me to https://github.com/dlang/druntime/blob/master/src/core/sys/posix/unistd.d#L2451 but declaring `vfork` here https://github.com/dlang/druntime/blob/master/src/core/sys/posix/unistd.d#L2479 (in the bootstrap compiler's druntime, of course) did not help, which leads me to believe the musl block is never evaluated thanks to D toolchain's poor support of anything other than x86_64 glibc.

Reply:
And yes, good point, I will not revert that line. I just sort of took a hammer approach with the revert, took the whole commit out.

Reply:
Gah, that's terrible.

> (in the bootstrap compiler's druntime, of course)

Are you sure that's correct? The bootstrap compiler should be building the new compiler with its own library and headers, right? Could be interesting to see what happens if you patch it in `dmd`.

Reply:
Also, do you know how they detect musl?

Reply:
>Are you sure that's correct?

I'm relatively certain, but I'll try patching `dmd`'s `druntime` anyways. Might work, who knows.

>Also, do you know how they detect musl?

Nope, no idea. I'll look into this tomorrow.

Reply:
> Nope, no idea. I'll look into this tomorrow.

It's not urgent, I was mostly wondering.

Also, moving to using `vfork` when it's not present for all headers is a big bug on their part.

Reply:
Tested patching `dmd`'s `druntime`, didn't help. Also tested adding a message pragma to the `version (CRuntime_Musl)` test in druntime - it did not log anything, which confirms that the branch is, indeed, never evaluated. Honestly not sure how this thing even builds (or works) at all.

Reply:
Ok, no problem. So just patch it to `fork()` for now, unless you want to explore deep into the build system :)

Reply:
Had a shot at tracking down where `CRuntime_Musl` is defined, couldn't find it; the dmd build toolchain (the entire codebase, really) is extremely convoluted. What I *do* know is that `CRuntime_Musl` is set when building both `dmd2.091` and `dmd` (`2.094.1`), made sure using a message pragma. Not sure why properly declaring `vfork()` doesn't help, it's really strange.